File: glyphbyte/pipeline.py
"""Image in, candidate byte sequences out, in whichever format the row was drawn.

Uncertainty is not hidden: every glyph keeps its ranked candidate bytes, and the
result lists the most probable whole sequences so a caller can query all of them
(cheap on nostr) and let a person pick the one that matches what they see.

Formats: the row is found once, then read as format 2 (icons and corner dots, the
default) and as format 1 (the first alphabet of rotated, filled pictograms). The
format whose model explains the glyphs better wins; when the two are close, the
other format's readings stay in the list as candidates, each tagged with its format.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import NamedTuple

import numpy as np

from . import v1
from .classify import Classifier, PatchScores, load_classifiers
from .detect import Detection, detect, rectify_frame
from .symbols import DEFAULT_FORMAT, describe

# how much more likely (nats, summed over the row) a turned reading must be before the
# glyphs overrule the underline; below this both readings are kept as candidates
ORIENT_MARGIN = 2.0
# format choice, fitted on 583 synthetic rows of both formats (2026-09-26): read format 1 when
# its log-likelihood per glyph beats format 2's by more than FORMAT_BIAS (95% of v1 rows and
# 96-97% of v2 rows go to the right format); keep the other format's readings as candidates
# while the difference is within FORMAT_FORK of that threshold
FORMAT_BIAS = 0.5
FORMAT_FORK = 0.5
RESCUE = 0.3          # the other format's model must be this sure a rejected candidate is a glyph


class Reading(NamedTuple):
    bytes: bytes
    p: float
    fmt: int


@dataclass
class SymbolRead:
    index: int
    candidates: list[tuple[int, float]]      # (byte, probability) best first
    frame_conf: float
    fmt: int = DEFAULT_FORMAT

    @property
    def byte(self) -> int:
        return self.candidates[0][0]

    @property
    def confidence(self) -> float:
        return self.candidates[0][1]

    def describe(self, byte: int | None = None) -> str:
        return describe(self.byte if byte is None else byte, self.fmt)


@dataclass
class Result:
    reads: list[SymbolRead]
    sequences: list[Reading]                 # best first; each tagged with its format
    warnings: list[str] = field(default_factory=list)
    detection: Detection | None = None
    fmt: int = DEFAULT_FORMAT                # the format the row was read as
    format_scores: dict = field(default_factory=dict)   # row log-likelihood per format that was tried

    @property
    def best(self) -> bytes:
        return self.sequences[0].bytes if self.sequences else b""

    def to_dict(self) -> dict:
        return {
            "best": self.best.hex(),
            "format": self.fmt,
            "sequences": [{"hex": s.bytes.hex(), "p": round(s.p, 4), "format": s.fmt} for s in self.sequences],
            "symbols": [
                {"index": r.index, "byte": r.byte, "glyph": r.describe(), "p": round(r.confidence, 4),
                 "candidates": [{"byte": b, "glyph": r.describe(b), "p": round(p, 4)} for b, p in r.candidates]}
                for r in self.reads
            ],
            "warnings": self.warnings,
        }


# ----------------------------------------------------------------------------- per glyph

def _dist_v2(sc: PatchScores) -> np.ndarray:
    """Joint probability over all 256 bytes for one format 2 cell: icon times each dot bit."""
    p_dots = np.ones(16)
    for n in range(16):
        for k in range(4):
            p_dots[n] *= sc.dots[k] if n >> (3 - k) & 1 else 1.0 - sc.dots[k]
    dist = np.outer(sc.icon, p_dots).ravel()     # index = icon << 4 | dots
    return dist / max(dist.sum(), 1e-12)


def _dist_v1(sc: PatchScores, frame_kind: int, frame_conf: float) -> np.ndarray:
    """Joint probability over all 256 bytes for one format 1 cell: pictogram and rotation,
    fill, and the frame's shape as the detector saw it."""
    p_frame = np.array([0.5, 0.5])
    p_frame[frame_kind] = 0.5 + 0.5 * frame_conf
    p_frame[1 - frame_kind] = 0.5 - 0.5 * frame_conf
    dist = (sc.sym_rot[:, None, None] * sc.fill[None, :, None] * p_frame[None, None, :]).ravel()
    return dist / max(dist.sum(), 1e-12)


def _candidates(dist: np.ndarray, fork_ratio: float, max_per_symbol: int) -> list[tuple[int, float]]:
    order = np.argsort(-dist)
    top = dist[order[0]]
    out = [(int(order[0]), float(top))]
    for b in order[1:max_per_symbol]:
        if dist[b] >= fork_ratio * top:
            out.append((int(b), float(dist[b])))
    return out


def _sequences(reads: list[SymbolRead], max_sequences: int) -> list[tuple[bytes, float]]:
    beam: list[tuple[list[int], float]] = [([], 1.0)]
    for r in reads:
        nxt = []
        for seq, p in beam:
            for b, pb in r.candidates:
                nxt.append((seq + [b], p * pb))
        nxt.sort(key=lambda t: -t[1])
        beam = nxt[:max_sequences]
    total = sum(p for _, p in beam) or 1.0
    return [(bytes(seq), p / total) for seq, p in beam]


def _row_ll(dists: list[np.ndarray]) -> float:
    """How well a format explains the row: the summed log probability of each glyph's best byte."""
    return float(sum(np.log(max(float(d.max()), 1e-12)) for d in dists))


# ----------------------------------------------------------------------------- per format

@dataclass
class _FormatRead:
    fmt: int
    reads: list[SymbolRead]
    sequences: list[Reading]
    ll: float
    warnings: list[str]
    frames: list = field(default_factory=list)
    patches: list = field(default_factory=list)
    up: np.ndarray | None = None
    d: np.ndarray | None = None


def _turn(up: np.ndarray, d: np.ndarray, k: int) -> tuple[np.ndarray, np.ndarray]:
    """True (up, d) when the glyphs in patches cut with (up, d) look turned k quarter turns
    counter-clockwise: k=1 means their tops point along -d."""
    for _ in range(k % 4):
        up, d = -d, up
    return up, d


def _cut(det: Detection, up, d, fmt: int):
    frames = sorted(det.frames, key=lambda f: float(f.center @ d))
    return frames, [rectify_frame(det.gray, f, d, up, det.light_ink, fmt) for f in frames]


def _read_v2(det: Detection, clf: Classifier, fork_ratio: float, max_per: int, max_seq: int) -> _FormatRead:
    warnings: list[str] = []
    frames, patches = det.frames, det.patches
    up, d = det.up, det.direction
    scores = clf.predict(patches)
    # every icon but box, plus and x has a top: the glyphs vote on which way the row is turned
    turns = (0, 2) if det.baseline is not None else (0, 1, 2, 3)
    ll = {k: float(sum(np.log(max(float(sc.orient[k]), 1e-6)) for sc in scores)) for k in turns}
    ranked = sorted(ll, key=lambda k: -ll[k])
    k_best = ranked[0]
    turned = None
    if k_best != 0 and ll[k_best] - ll[0] > ORIENT_MARGIN:
        up, d = _turn(det.up, det.direction, k_best)
        frames, patches = _cut(det, up, d, 2)
        scores = clf.predict(patches)
        warnings.append(f"the glyphs read turned {90 * k_best} degrees from the underline's side; turned the row")
    elif k_best != 0 or ll[ranked[0]] - ll[ranked[1]] < ORIENT_MARGIN:
        turned = ranked[1] if k_best == 0 else k_best
    dists = [_dist_v2(sc) for sc in scores]
    reads = [SymbolRead(index=i, candidates=_candidates(dd, fork_ratio, max_per), frame_conf=f.quad_ratio, fmt=2)
             for i, (f, dd) in enumerate(zip(frames, dists))]
    seqs = [Reading(b, p, 2) for b, p in _sequences(reads, max_seq)]
    if turned is not None and not det.start_known:
        up2, d2 = _turn(det.up, det.direction, turned)
        f2, p2 = _cut(det, up2, d2, 2)
        reads2 = [SymbolRead(index=i, candidates=_candidates(_dist_v2(sc), fork_ratio, max_per), frame_conf=f.quad_ratio, fmt=2)
                  for i, (f, sc) in enumerate(zip(f2, clf.predict(p2)))]
        w = 1.0 / (1.0 + np.exp(ll[0] - ll[turned]))
        seqs = sorted([Reading(b, p * (1 - w), 2) for b, p, _ in seqs] +
                      [Reading(b, p * w, 2) for b, p in _sequences(reads2, max_seq)], key=lambda s: -s.p)[:max_seq]
        warnings.append(f"not sure which way is up: candidates include the row turned {90 * turned} degrees")
    return _FormatRead(2, reads, seqs, _row_ll(dists), warnings, frames, patches, up, d)


def _read_v1(det: Detection, clf: Classifier, fork_ratio: float, max_per: int, max_seq: int) -> _FormatRead:
    warnings: list[str] = []
    frames, patches = _cut(det, det.up, det.direction, 1)
    dists = [_dist_v1(sc, f.kind, f.quad_ratio) for sc, f in zip(clf.predict(patches), frames)]
    reads = [SymbolRead(index=i, candidates=_candidates(dd, fork_ratio, max_per), frame_conf=f.quad_ratio, fmt=1)
             for i, (f, dd) in enumerate(zip(frames, dists))]
    seqs = [Reading(b, p, 1) for b, p in _sequences(reads, max_seq)]
    if not det.start_known:
        # v1 pictograms are drawn in every rotation, so they cannot say which way is up: the row
        # turned 180 degrees reads in reverse, every glyph's rotation two quarter turns on
        rev = [Reading(bytes(v1.turned(b, 2) for b in reversed(s.bytes)), s.p * 0.5, 1) for s in seqs]
        seqs = sorted([Reading(s.bytes, s.p * 0.5, 1) for s in seqs] + rev, key=lambda s: -s.p)[:max_seq]
        warnings.append("format 1 has no top to read: candidates include the row turned 180 degrees")
    return _FormatRead(1, reads, seqs, _row_ll(dists), warnings, frames, patches, det.up, det.direction)


# ----------------------------------------------------------------------------- entry

_DEFAULT: dict[int, Classifier] | None = None


def _classifiers(classifier) -> dict[int, Classifier]:
    global _DEFAULT
    if isinstance(classifier, dict):
        return classifier
    if isinstance(classifier, Classifier):
        return {classifier.fmt: classifier}
    if _DEFAULT is None:
        _DEFAULT = load_classifiers()
    return _DEFAULT


def _junk_fn(clfs: dict[int, Classifier]):
    """Probability that a candidate frame holds no glyph of any format, in any rotation. Format 2
    decides; format 1 gets a second look only at what format 2 rejects."""
    up, d = np.array([0.0, -1.0]), np.array([1.0, 0.0])

    def fn(gray, frames):
        fmts = sorted(clfs, reverse=True)
        pj = clfs[fmts[0]].junk_probabilities([rectify_frame(gray, f, d, up, f.light_ink, fmts[0]) for f in frames])
        for fmt in fmts[1:]:
            idx = [i for i, p in enumerate(pj) if p > 0.5]
            if idx:
                p2 = clfs[fmt].junk_probabilities([rectify_frame(gray, frames[i], d, up, frames[i].light_ink, fmt) for i in idx])
                for i, p in zip(idx, p2):
                    if p < RESCUE:          # only a clear glyph of the other format overrules
                        pj[i] = min(pj[i], p)
        return pj
    return fn


def read_image(image: np.ndarray, classifier: Classifier | dict[int, Classifier] | None = None,
               max_sequences: int = 8, fork_ratio: float = 0.2, max_per_symbol: int = 4,
               fmt: int | str = "auto") -> Result:
    """Read a photo. `classifier` is one Classifier (its format only), a dict {format:
    Classifier}, or None for the bundled models. `fmt` is "auto" (both formats, the better
    one wins), 1 or 2."""
    clfs = _classifiers(classifier)
    if fmt != "auto":
        if int(fmt) not in clfs:
            clfs = {**clfs, **load_classifiers(formats=(int(fmt),))}
        clfs = {int(fmt): clfs[int(fmt)]}
    det = detect(image, junk_fn=_junk_fn(clfs))
    warnings = list(det.warnings)
    if not det.frames:
        return Result(reads=[], sequences=[], warnings=warnings + ["no glyphs found"], detection=det,
                      fmt=max(clfs))
    readers = {2: _read_v2, 1: _read_v1}
    results = {}
    for f in sorted(clfs, reverse=True):
        # format 1 wins only by beating format 2 by FORMAT_BIAS per glyph, and a log-likelihood is at
        # most 0: when format 2 is above -FORMAT_BIAS per glyph, format 1 cannot win and is not read
        if f == 1 and 2 in results and results[2].ll / max(1, len(det.frames)) > -FORMAT_BIAS:
            continue
        results[f] = readers[f](det, clfs[f], fork_ratio, max_per_symbol, max_sequences)
    if len(results) == 1:
        (chosen,) = results.values()
        sequences = chosen.sequences
    else:
        n = max(1, len(det.frames))
        margin = (results[1].ll - results[2].ll) / n - FORMAT_BIAS  # per glyph; > 0 favours format 1
        chosen = results[1] if margin > 0 else results[2]
        other = results[2] if margin > 0 else results[1]
        sequences = chosen.sequences
        if abs(margin) < FORMAT_FORK:
            w = 1.0 / (1.0 + np.exp(abs(margin) * n))                # the other format's share
            sequences = sorted([Reading(s.bytes, s.p * (1 - w), s.fmt) for s in chosen.sequences] +
                               [Reading(s.bytes, s.p * w, s.fmt) for s in other.sequences],
                               key=lambda s: -s.p)[:max_sequences]
            warnings.append(f"the row could be format {other.fmt}: its readings are among the candidates")
        if chosen.fmt == 1:
            warnings.append("read as format 1, the first glyphbyte alphabet")
    warnings += chosen.warnings
    det.frames, det.patches, det.up, det.direction = chosen.frames, chosen.patches, chosen.up, chosen.d
    return Result(reads=chosen.reads, sequences=sequences, warnings=warnings, detection=det, fmt=chosen.fmt,
                  format_scores={f: r.ll for f, r in results.items()})

File: napplet/src/pipeline.js
// Image to candidate byte sequences, in whichever format the row was drawn: port of glyphbyte/pipeline.py.
// The row is found once, then read as format 2 (icons and corner dots, the default) and as format 1 (the
// first alphabet of rotated, filled pictograms). The better-explained format wins; when the two are close,
// the other format's readings stay among the candidates, each tagged with its format.
import { detect, rectifyFrame } from './detect.js';
import { classify, modelFormat } from './nn.js';

export const SYMBOLS = ['house', 'heart', 'drop', 'moon', 'crown', 'arrow', 'box', 'triangle', 'pie', 'tree', 'plus', 'flag', 'x', 'bolt', 'star', 'fish'];
export const SYMBOLS_V1 = ['house', 'chevron', 'bookmark', 'crown', 'drop', 'tee', 'u', 'mountain', 'arrow', 'heart', 'crescent', 'cloud', 'snowman', 'l', 'trapezoid', 'pacman'];
export const DOT_NAMES = ['top-left', 'top-right', 'bottom-right', 'bottom-left'];   // bit 3, 2, 1, 0
export const DEFAULT_FORMAT = 2;

export function unpack(b, fmt = DEFAULT_FORMAT) {
  if (fmt === 1) return { fmt: 1, symbol: b >> 4, rotation: (b >> 2) & 3, fill: (b >> 1) & 1, frame: b & 1, name: SYMBOLS_V1[b >> 4] };
  const dots = b & 15; return { fmt: 2, icon: b >> 4, dots, name: SYMBOLS[b >> 4], corners: DOT_NAMES.filter((_, i) => dots >> (3 - i) & 1) };
}
export function describe(b, fmt = DEFAULT_FORMAT) {
  const g = unpack(b, fmt);
  if (fmt === 1) return `${g.name} rotated ${g.rotation * 90} deg, ${g.fill ? 'filled' : 'outline'}, ${g.frame ? 'circle' : 'square'} frame`;
  const c = g.corners; return `${g.name}, ` + (c.length ? (c.length > 1 ? 'dots ' : 'dot ') + c.join(', ') : 'no dots');
}
// the byte a format 1 glyph reads as when the whole row is turned clockwise by quarter turns
export function turnedV1(b, q) { return (b & 0xf3) | ((((b >> 2) & 3) + q) % 4) << 2; }

// how much more likely (nats, summed over the row) a turned reading must be before the glyphs overrule the
// underline; below this both readings are kept. Format choice constants as in pipeline.py.
const ORIENT_MARGIN = 2.0, FORMAT_BIAS = 0.5, FORMAT_FORK = 0.5, RESCUE = 0.3;   // the format ones are per glyph

function distV2(sc) {
  const dist = new Float64Array(256); let s = 0;
  for (let n = 0; n < 16; n++) {
    let pd = 1; for (let k = 0; k < 4; k++) pd *= (n >> (3 - k) & 1) ? sc.dots[k] : 1 - sc.dots[k];
    for (let i = 0; i < 16; i++) { const v = sc.icon[i] * pd; dist[(i << 4) | n] = v; s += v; }
  }
  for (let i = 0; i < 256; i++) dist[i] /= s || 1;
  return dist;
}
function distV1(sc, kind, conf) {
  const pf = [0.5, 0.5]; pf[kind] = 0.5 + 0.5 * conf; pf[1 - kind] = 0.5 - 0.5 * conf;
  const dist = new Float64Array(256); let s = 0;
  for (let sr = 0; sr < 64; sr++) for (let fill = 0; fill < 2; fill++) for (let fr = 0; fr < 2; fr++) { const v = sc.symRot[sr] * sc.fill[fill] * pf[fr]; dist[(sr << 2) | (fill << 1) | fr] = v; s += v; }
  for (let i = 0; i < 256; i++) dist[i] /= s || 1;
  return dist;
}
function candidates(dist, forkRatio, maxPer) {
  const order = Array.from(dist.keys()).sort((a, b) => dist[b] - dist[a]), top = dist[order[0]], out = [[order[0], top]];
  for (let i = 1; i < maxPer; i++) if (dist[order[i]] >= forkRatio * top) out.push([order[i], dist[order[i]]]);
  return out;
}
const hexOf = seq => seq.map(b => b.toString(16).padStart(2, '0')).join('');
function sequences(reads, maxSeq, fmt) {
  let beam = [[[], 1]];
  for (const r of reads) { const nxt = []; for (const [seq, p] of beam) for (const [b, pb] of r.candidates) nxt.push([seq.concat([b]), p * pb]); nxt.sort((a, b) => b[1] - a[1]); beam = nxt.slice(0, maxSeq); }
  const tot = beam.reduce((s, x) => s + x[1], 0) || 1;
  return beam.map(([seq, p]) => ({ bytes: seq, hex: hexOf(seq), p: p / tot, format: fmt }));
}
const rowLL = dists => dists.reduce((s, d) => s + Math.log(Math.max(Math.max(...d), 1e-12)), 0);
const merge = (a, wa, b, wb, maxSeq) => a.map(s => ({ ...s, p: s.p * wa })).concat(b.map(s => ({ ...s, p: s.p * wb }))).sort((x, y) => y.p - x.p).slice(0, maxSeq);

// true [up, d] when glyphs cut with (up, d) look turned k quarter turns counter-clockwise
function turn(up, d, k) { for (let i = 0; i < ((k % 4) + 4) % 4; i++) [up, d] = [[-d[0], -d[1]], up]; return [up, d]; }

function cut(det, up, d, fmt) {
  const frames = det.frames.slice().sort((a, b) => (a.center[0] * d[0] + a.center[1] * d[1]) - (b.center[0] * d[0] + b.center[1] * d[1]));
  return { frames, patches: frames.map(f => rectifyFrame(det.gray, det.W, det.H, f, d, up, det.lightInk, fmt)) };
}
function readsOf(frames, dists, forkRatio, maxPer, fmt) {
  return frames.map((f, i) => { const c = candidates(dists[i], forkRatio, maxPer); return { index: i, candidates: c, byte: c[0][0], p: c[0][1], frameConf: f.conf, format: fmt }; });
}

function readV2(det, model, o) {
  const warnings = [];
  let { frames, up, direction: d } = det, patches = det.patches.length === frames.length ? det.patches : cut(det, up, d, 2).patches;
  let scores = patches.map(p => classify(model, p, 64));
  // every icon but box, plus and x has a top: the glyphs vote on which way the row is turned
  const turns = det.baseline ? [0, 2] : [0, 1, 2, 3];
  const ll = {}; for (const k of turns) ll[k] = scores.reduce((s, sc) => s + Math.log(Math.max(sc.orient[k], 1e-6)), 0);
  const ranked = turns.slice().sort((a, b) => ll[b] - ll[a]), kBest = ranked[0];
  let turned = null;
  if (kBest !== 0 && ll[kBest] - ll[0] > ORIENT_MARGIN) {
    [up, d] = turn(det.up, det.direction, kBest); ({ frames, patches } = cut(det, up, d, 2)); scores = patches.map(p => classify(model, p, 64));
    warnings.push(`the glyphs read turned ${90 * kBest} degrees from the underline's side; turned the row`);
  } else if (kBest !== 0 || ll[ranked[0]] - ll[ranked[1]] < ORIENT_MARGIN) turned = kBest === 0 ? ranked[1] : kBest;
  const dists = scores.map(distV2), reads = readsOf(frames, dists, o.forkRatio, o.maxPer, 2);
  let seqs = sequences(reads, o.maxSeq, 2);
  if (turned !== null && !det.startKnown) {
    const [u2, d2] = turn(det.up, det.direction, turned), c2 = cut(det, u2, d2, 2);
    const seq2 = sequences(readsOf(c2.frames, c2.patches.map(p => distV2(classify(model, p, 64))), o.forkRatio, o.maxPer, 2), o.maxSeq, 2);
    const w = 1 / (1 + Math.exp(ll[0] - ll[turned]));
    seqs = merge(seqs, 1 - w, seq2, w, o.maxSeq);
    warnings.push(`not sure which way is up: candidates include the row turned ${90 * turned} degrees`);
  }
  return { fmt: 2, reads, sequences: seqs, ll: rowLL(dists), warnings, frames, patches, up, d };
}

function readV1(det, model, o) {
  const warnings = [], { frames, patches } = cut(det, det.up, det.direction, 1);
  const dists = patches.map((p, i) => distV1(classify(model, p, 64), frames[i].kind, frames[i].conf));
  const reads = readsOf(frames, dists, o.forkRatio, o.maxPer, 1);
  let seqs = sequences(reads, o.maxSeq, 1);
  if (!det.startKnown) {
    // v1 pictograms are drawn in every rotation, so they cannot say which way is up: the row turned
    // 180 degrees reads in reverse, every glyph's rotation two quarter turns on
    const rev = seqs.map(s => { const b = s.bytes.slice().reverse().map(x => turnedV1(x, 2)); return { bytes: b, hex: hexOf(b), p: s.p, format: 1 }; });
    seqs = merge(seqs, 0.5, rev, 0.5, o.maxSeq);
    warnings.push('format 1 has no top to read: candidates include the row turned 180 degrees');
  }
  return { fmt: 1, reads, sequences: seqs, ll: rowLL(dists), warnings, frames, patches, up: det.up, d: det.direction };
}

// models: one loaded model (its format only) or {1: model, 2: model}. opts.format: 'auto' (default), 1 or 2.
export function readImage(gray, W, H, models, opts = {}) {
  const o = { maxSeq: opts.maxSequences || 8, forkRatio: opts.forkRatio || 0.2, maxPer: opts.maxPerSymbol || 4 };
  let byFmt = models && models.w ? { [modelFormat(models)]: models } : { ...models };
  if (opts.format && opts.format !== 'auto') byFmt = { [opts.format]: byFmt[opts.format] };
  const fmts = Object.keys(byFmt).map(Number).filter(f => byFmt[f]).sort((a, b) => b - a);
  // junk: format 2 decides; format 1 gets a second look only at what format 2 rejects
  const junkFn = (g, w, h, frames) => {
    const pj = frames.map(f => classify(byFmt[fmts[0]], rectifyFrame(g, w, h, f, [1, 0], [0, -1], f.lightInk, fmts[0]), 64).junk);
    // only a clear glyph of the other format overrules (RESCUE as in pipeline.py)
    for (const fm of fmts.slice(1)) frames.forEach((f, i) => { if (pj[i] > 0.5) { const p = classify(byFmt[fm], rectifyFrame(g, w, h, f, [1, 0], [0, -1], f.lightInk, fm), 64).junk; if (p < RESCUE) pj[i] = Math.min(pj[i], p); } });
    return pj;
  };
  const det = detect(gray, W, H, junkFn);
  const warnings = det.warnings.slice();
  if (!det.frames.length) return { reads: [], sequences: [], format: fmts[0], warnings: warnings.concat(['no glyphs found']), detection: det };
  // format 1 wins only by beating format 2 by FORMAT_BIAS per glyph and scores at most 0, so it is not
  // read when format 2 is above -FORMAT_BIAS per glyph (fmts is sorted 2 before 1)
  const results = {};
  for (const f of fmts) {
    if (f === 1 && results[2] && results[2].ll / Math.max(1, det.frames.length) > -FORMAT_BIAS) continue;
    results[f] = (f === 1 ? readV1 : readV2)(det, byFmt[f], o);
  }
  let chosen, seqs;
  if (!(results[1] && results[2])) { chosen = results[2] || results[1]; seqs = chosen.sequences; }
  else {
    const n = Math.max(1, det.frames.length), margin = (results[1].ll - results[2].ll) / n - FORMAT_BIAS;   // per glyph; > 0 favours format 1
    chosen = margin > 0 ? results[1] : results[2]; const other = margin > 0 ? results[2] : results[1];
    seqs = chosen.sequences;
    if (Math.abs(margin) < FORMAT_FORK) {
      const w = 1 / (1 + Math.exp(Math.abs(margin) * n));
      seqs = merge(chosen.sequences, 1 - w, other.sequences, w, o.maxSeq);
      warnings.push(`the row could be format ${other.fmt}: its readings are among the candidates`);
    }
    if (chosen.fmt === 1) warnings.push('read as format 1, the first glyphbyte alphabet');
  }
  Object.assign(det, { frames: chosen.frames, patches: chosen.patches, up: chosen.up, direction: chosen.d });
  return { reads: chosen.reads, sequences: seqs, best: seqs[0].hex, format: chosen.fmt, warnings: warnings.concat(chosen.warnings), detection: det };
}
